beluga_ros/tools/conversion_utils.py

```diff
@@ -0,0 +1,261 @@
+
+    def __eq__(self, other: "NormalDistribution") -> bool:
+        """Element-wise equality."""
+        return np.allclose(self.mean, other.mean) and np.allclose(
+            self.covariance, other.covariance
+        )
+
+
+class NDTMap:
+    def __init__(self, resolution: float) -> None:
+        """Create a new NDT map with cell size of 'resolution'."""
+        self._resolution = resolution
+        self._grid: Dict[DiscreteCell, NormalDistribution] = {}
+
+    def add_distribution(self, cell: DiscreteCell, ndt: NormalDistribution):
+        """Add a new cell with its distribution."""
+        self._grid[cell] = ndt
+
+    def _get_plot_bounds(self):
+        min_x = float("inf")
+        min_y = float("inf")
+
+        max_x = -float("inf")
+        max_y = -float("inf")
+        for discrete_cell in self._grid.keys():
+            x = discrete_cell.x * self._resolution + self._resolution / 2
+            y = discrete_cell.y * self._resolution + self._resolution / 2
+            min_x = min(min_x, x)
+            max_x = max(max_x, x)
+
+            min_y = min(min_y, y)
+            max_y = max(max_y, y)
+
+        step_x = (max_x - min_x) / 100.0
+        step_y = (max_y - min_y) / 100.0
+
+        PADDING = 1  # [m]
+
+        min_x -= PADDING
+        max_x += PADDING
+        min_y -= PADDING
+        max_y += PADDING
+
+        xx, yy = np.mgrid[min_x:max_x:step_x, min_y:step_y]
+        return (xx, yy, np.dstack((xx, yy)))
+
+    def __eq__(self, other: "NDTMap") -> bool:
+        """Equality for two NDT maps."""
+        return self._grid == other._grid
+
+    def plot(self, show: bool = False) -> plt.figure:
+        """
+        Plot the map using contour plots into the current figure and returns it.
+
+        Optionally show the plot based on the 'show' parameter.
+        """
+        xx, yy, pos = self._get_plot_bounds()
+
+        for ndt in self._grid.values():
+            plt.contour(
+                xx,
+                yy,
+                ndt.get_scipy_representation().pdf(pos),
+                [0.1, 0.5],
+                alpha=0.2,
+                linewidths=3,
+                extend="max",
+            )
+        if show:
+            plt.show()
+        return plt.gcf()
+
+    def serialize_to(self, output_file_path: Path):
+        """
+        Serialize the NDT map into an HDF5 format.
+
+        See https://docs.hdfgroup.org/hdf5/develop/_intro_h_d_f5.html for details on the format.
+        It'll create 4 datasets:
+            - "resolution": () resolution for the discrete grid (cells are resolution x
+              resolution m^2 squares).
+            - "discrete_cells": (NUM_CELLS, 2) that contains the discrete cell coordinates.
+            - "means": (NUM_CELLS, 2) that contains the 2d mean of the normal distribution
+              of the cell.
+            - "covariances": (NUM_CELLS, 2, 2) Contains the covariance for each cell.
```

Comment:
meta: Wouldn't it be more convenient to have all these in a single dataset?

Reply:
Not completely sure, to be honest.
In my mind this'll make C++ parsing easier, but I'd vote for keeping it this way, and change it if when working on the C++ parsing side of things it makes more sense.

Reply:
Not sure TBH.

But in my mind splitting them should make it easier to parse from C++ land.
I proppose we move forward with this version and I'll revisit when I get to C++ side of this. WDYT?

Reply:
Sounds good!